DataFormats/TrackSoA/interface/TracksSoA.h

```diff
@@ -12,91 +12,141 @@
+                                                         uint16_t maxLayers,
+                                                         uint32_t const *__restrict__ layerStarts,
+                                                         int32_t i) {
+    auto start = (i == 0) ? 0 : tracks[i - 1].hitOffsets();
+    auto end = tracks[i].hitOffsets();
+    auto hitId = hits[start].id();
+    int nl = 1;
+    auto ol = 0;
+    while (hitId >= layerStarts[ol + 1] and ol < maxLayers)
+      ++ol;
+    ++start;
+    for (; start < end; ++start) {
+      hitId = hits[start].id();
+      auto il = 0;
+      while (hitId >= layerStarts[il + 1] and il < maxLayers)
+        ++il;
+      if (il != ol)
+        ++nl;
+      ol = il;
+    }
+    return nl;
+  }
+
+  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE float charge(const TrackSoAConstView &tracks, int32_t i) {
     //was: std::copysign(1.f, tracks[i].state()(2)). Will be constexpr with C++23
     float v = tracks[i].state()(2);
     return float((0.0f < v) - (v < 0.0f));
   }
 
-  template <typename ConstView, typename = std::enable_if_t<isTrackSoAConstView<ConstView>>>
-  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE constexpr float phi(ConstView const& tracks, int32_t i) {
+  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE float phi(const TrackSoAConstView &tracks, int32_t i) {
     return tracks[i].state()(0);
   }
 
-  template <typename ConstView, typename = std::enable_if_t<isTrackSoAConstView<ConstView>>>
-  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE constexpr float tip(ConstView const& tracks, int32_t i) {
+  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE float tip(const TrackSoAConstView &tracks, int32_t i) {
     return tracks[i].state()(1);
   }
 
-  template <typename ConstView, typename = std::enable_if_t<isTrackSoAConstView<ConstView>>>
-  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE constexpr float zip(ConstView const& tracks, int32_t i) {
+  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE float zip(const TrackSoAConstView &tracks, int32_t i) {
     return tracks[i].state()(4);
   }
 
-  template <typename ConstView, typename = std::enable_if_t<isTrackSoAConstView<ConstView>>>
-  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE constexpr bool isTriplet(ConstView const& tracks, int32_t i) {
+  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE bool isTriplet(const TrackSoAConstView &tracks, int32_t i) {
     return tracks[i].nLayers() == 3;
   }
 
+  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE static int nHits(const TrackSoAConstView &tracks, int i) {
+    auto start = (i == 0) ? 0 : tracks[i - 1].hitOffsets();
+    return tracks[i].hitOffsets() - start;
+  }
+
+  // state at the beam spot: { phi, tip, 1/pt, cotan(theta), zip }
+
+  // variable of non-literal type 'MapType'
+  // (aka 'Map<Eigen::Matrix<float, 15, 1, 0, 15, 1>, 0, Eigen::InnerStride<Eigen::Dynamic>>')
+  // cannot be defined in a function before C++2b
+  template <typename V3, typename M3, typename V2, typename M2>
+  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE static void copyFromCircle(
```

Comment:
Also this function should not be `static`

Reply:
Fixing these, thanks! 
(These are all remnants of the fact that these were in a struct before)